ESP-IDF support

fedepell opened the pull request:
Adds support for serial for ESP-IDF and introduces a document on how to use it in that environment.

Assumes to be merged after #741 (I will rebase accordingly in case).

fedepell committed:
Add support for ESP-IDF platform complete with documentation

diplfranzhoepfinger commented:
sorry, i overlooked this !

diplfranzhoepfinger commented on src/modbus-rtu.c:
```diff
@@ -26,6 +26,13 @@
 #endif
 
+
+#if ESP_IDF
```
why here 

```
#if ESP_IDF
```

and not

```
#if defined(ESP_PLATFORM)
```

fedepell replied:
yeah on the naming (`ESP_IDF` vs `ESP_PLATFORM`) we can put what @stephane think it's best!

stephane replied:
I have a preference for `ESP_PLATFORM`

diplfranzhoepfinger replied:
Me too

fedepell replied:
Changed to `ESP_PLATFORM`!

diplfranzhoepfinger commented on src/modbus-rtu.c:
```diff
@@ -503,6 +510,90 @@ static int _modbus_rtu_connect(modbus_t *ctx)
+// EspressIf SDK
+static int _modbus_rtu_connect(modbus_t *ctx)
+{
+    modbus_rtu_t *ctx_rtu = ctx->backend_data;
```
This is nice in my Opinion. 

this i meant when i wrote in #767

diplfranzhoepfinger commented on src/modbus-rtu.c:
```diff
@@ -503,6 +510,90 @@ static int _modbus_rtu_connect(modbus_t *ctx)
+
+    return 0;
+}
+
```
where do you set the Pin Numbers ? 

```c
    // Set UART pins
	uart_set_pin(UART_NUM_1, 13, 1, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE);
```

https://github.com/Fliegl-Agrartechnik-GmbH/libmodbus-examples-esp-idf/blob/90839c84e529693d58e2acf53a550c82bba62d83/modbus-rtu/main/main.c#L59

fedepell replied:
yeah you saw then later down, but my goal was to keep the library as clean as possible and leave all the hardware dependant parts outside. This makes the MR less intrusive as possible. So I deferred all that to before the library calls as per README.

diplfranzhoepfinger commented on src/modbus-rtu.c:
```diff
@@ -503,6 +510,90 @@ static int _modbus_rtu_connect(modbus_t *ctx)
+        return -1;
+    }
+
+    switch (ctx_rtu->stop_bit) {
```
nice. like it.

diplfranzhoepfinger commented on src/modbus-rtu.c:
```diff
@@ -503,6 +510,90 @@ static int _modbus_rtu_connect(modbus_t *ctx)
+    };
+
+    // Configure UART parameters
+    ESP_ERROR_CHECK(uart_param_config(uart_num, &uart_config));
```
also where do you set the right HW Config ??? 

*   Circuit A: Collision Detection Circuit https://docs.espressif.com/projects/esp-idf/en/release-v5.3/esp32s3/api-reference/peripherals/uart.html#circuit-a-collision-detection-circuit

*   Circuit B: Manual Switching Transmitter/Receiver Without Collision Detection https://docs.espressif.com/projects/esp-idf/en/release-v5.3/esp32s3/api-reference/peripherals/uart.html#circuit-b-manual-switching-transmitter-receiver-without-collision-detection

*   Circuit C: Auto Switching Transmitter/Receiver https://docs.espressif.com/projects/esp-idf/en/release-v5.3/esp32s3/api-reference/peripherals/uart.html#circuit-c-auto-switching-transmitter-receiver


it normally requires something like this:

fedepell replied:
:+1:  (down below ;) )

diplfranzhoepfinger commented on src/esp-idf/README.md:
```diff
@@ -0,0 +1,54 @@
+If you desire to use the library for serial communication, you will need to do a few
+more hardware configuration steps before using the `modbus_new_rtu` call, namely:
+
+- Configure, if needed, any pins for the used uart via `uart_set_pin`
```
ah, here.

diplfranzhoepfinger commented on src/esp-idf/README.md:
```diff
@@ -0,0 +1,54 @@
+
+- Configure, if needed, any pins for the used uart via `uart_set_pin`
+
+- Install the uart driver via the `uart_driver_install`
```
yes

diplfranzhoepfinger commented on src/esp-idf/README.md:
```diff
@@ -0,0 +1,54 @@
+
+- Install the uart driver via the `uart_driver_install`
+
+- Configure, if needed, the uart mode (ie. set it to half duplex) via `uart_set_mode`
```
OK. clear. makes sense. 

good Job.

fedepell replied:
Thanks :)

fedepell commented:
> sorry, i overlooked this !

No problem :) Sorry for my delay now, was on a short vacation ;)

I hope we can get this (ESP support in general, in some variant) in as it may be very handy to have it mainstreamed. Also I've done a lot of testing with my code (both serial and TCP) and works like a charm (while the official internal library has a lot of limitations, ie. trivially missing possibility to have multiple connections which is just addressed in the beta).

robin-twice commented:
@fedepell Thanks for your efforts with the ESP-IDF platform support!

I've tried to use your implementation as well, but stumbled upon some issues and wanted to asked whether you can confirm that: Using the ESP-IDF UART VFS with Modbus RTU, the `uart_read()` function does return early on some special characters (`0x0A`, `0x0D`) depending on the line endings, which then creates either CRC errors or incomplete messages thereafter:

```
2024-10-07 07:23:09 Waiting for a confirmation...
2024-10-07 07:23:09 <47><03><02><00><0D><0A><B1>
2024-10-07 07:23:09 ERROR CRC received 0xB10A != CRC calculated 0x4EF0
2024-10-07 07:23:10 [47][03][9C][98][00][01][25][13]
2024-10-07 07:23:10 Waiting for a confirmation...
2024-10-07 07:23:10 <8C><47><03><02><00>
2024-10-07 07:23:10 Request for slave 140 ignored (not 71)
2024-10-07 07:23:10 The responding slave 140 isn't the requested slave 71
```

I've found a ticket on the ESP-IDF project where this issue with  VFS UART read() behaviour is being addressed:
https://github.com/espressif/esp-idf/issues/14155

Would be interested in your opinion about that and whether you've seen similar issues.

stephane commented:
Could you rebase your draft PR against master branch, please?

fedepell commented:
> Could you rebase your draft PR against master branch, please?

Rebased on master and squashed commits to one!

fedepell commented:
@robin-twice : sorry for the late response, very busy period :( 

I've noticed some seldom "bad crc" messages, but didn't investigate in big detail as I saw (I did many long term tests and kept statistics of errors by type and so on) the same (very low) amount, actually even slightly higher, using the ESP native modbus library (which _may_ suffer from the same problem if it reads data same way?) which I compared to (so I assumed it was just my lab-grade cabling or the 485 level adapter I have flying).

I'll try to give it a deeper look ASAP, hopefully this weekend or next week!

fedepell commented:
I just pushed a few moments ago two little changes:
- use `uart_vfs_dev_use_driver` instead of `esp_vfs_dev_uart_use_driver` as the latter is deprecated and resulted in a warning
- Add in the docs also a note of having possibly (if console UART is disabled, something I faced today ;-) ) to explicitly initialize the VFS in your code if needed (I would not put this inside libmodbus as it is a system wide decision) with `uart_vfs_dev_register` (see: https://github.com/stephane/libmodbus/blob/a4eb887b6cb1926efde97ff6c8db913429c11986/src/esp-idf/README.md?plain=1#L54)

fedepell committed:
modbus-rtu: esp-idf port, make sure line endings are ignored on UART

By default some line ending conversion may be done, so make sure to
disable this on the UART used with libmodbus.

fedepell commented:
> @fedepell Thanks for your efforts with the ESP-IDF platform support!
> 
> I've tried to use your implementation as well, but stumbled upon some issues and wanted to asked whether you can confirm that: Using the ESP-IDF UART VFS with Modbus RTU, the `uart_read()` function does return early on some special characters (`0x0A`, `0x0D`) depending on the line endings, which then creates either CRC errors or incomplete messages thereafter:
> 
> ```
> 2024-10-07 07:23:09 Waiting for a confirmation...
> 2024-10-07 07:23:09 <47><03><02><00><0D><0A><B1>
> 2024-10-07 07:23:09 ERROR CRC received 0xB10A != CRC calculated 0x4EF0
> 2024-10-07 07:23:10 [47][03][9C][98][00][01][25][13]
> 2024-10-07 07:23:10 Waiting for a confirmation...
> 2024-10-07 07:23:10 <8C><47><03><02><00>
> 2024-10-07 07:23:10 Request for slave 140 ignored (not 71)
> 2024-10-07 07:23:10 The responding slave 140 isn't the requested slave 71
> ```
> 
> I've found a ticket on the ESP-IDF project where this issue with VFS UART read() behaviour is being addressed: [espressif/esp-idf#14155](https://github.com/espressif/esp-idf/issues/14155)
> 
> Would be interested in your opinion about that and whether you've seen similar issues.

Sorry it took forever to me to test :-( (`$DAYJOB` priorities ;-) )

I partially confirm what you found: to reproduce I crafted requests of 10 registers (so a `0x0a` gets there with no doubts) and could reproduce, the write was not going through.

I've then added (see last pushed commit) the `ESP_LINE_ENDING_LF` configuration (to both read and write) and that seems to be gone now. And now I don't get also that seldom errors I saw (interesting enough the same problem was in the ESP internal modbus protocol, so that mislead me :-( ).

But as I see here in libmodbus we are anyway `select()`-ing and `read()`-ing until a certain length we want (or timeout of cours) so we are anyhow looping on it (as in "alternatives you have considered" section in the ticket you linked to), so I'm not sure the problem should apply?

If it should, it seems that the solution was finally merged in master (https://github.com/espressif/esp-idf/commit/92b4c62d06cf10fc731b260ffee659a904bf0860#diff-8f70148cf331b8d8293e6c953cf8d7ce6a5951cb0ceb6c087ffd9c176aabc292R172) so should be out with next release, so maybe still not worth doing further workarounds? (what do you say @stephane ?)

Many thanks for your feedback and sorry again for the very late feedback :(

robin-twice commented:
@fedepell No worries, thanks for coming back to it :)
`ESP_LINE_ENDING_LF` did help somewhat for me, but there were still some cases where it returned early (can't remember exactly whether on `0x0A` or `0x0D`).

Anyway, I also agree that with https://github.com/espressif/esp-idf/issues/14155 fixed it should not be an issue anymore, I haven't tested it myself though. Maybe you could also add to the readme that it needs >= xx.xx release of ESP-IDF to properly work.

fedepell commented:
> @fedepell No worries, thanks for coming back to it :) `ESP_LINE_ENDING_LF` did help somewhat for me, but there were still some cases where it returned early (can't remember exactly whether on `0x0A` or `0x0D`).
> 
> Anyway, I also agree that with [espressif/esp-idf#14155](https://github.com/espressif/esp-idf/issues/14155) fixed it should not be an issue anymore, I haven't tested it myself though. Maybe you could also add to the readme that it needs >= xx.xx release of ESP-IDF to properly work.

Added a few lines to address this heads-up and a reference to the ticket, good to save possible headaches to some future user :)
